ocaml/xapi/test_guest_agent.ml

```diff
@@ -0,0 +1,371 @@
+      let string_of_output_t = Test_printers.(assoc_list string string)
+    end
+
+    type 'a mtree = 
+      | Lf of 'a * 'a
+      | Mt of 'a * 'a mtree list
+
+    let has_name name = function
+      | Lf (n, _) -> n = name
+      | Mt (n, _) -> n = name
+
+    let get_name = function
+      | Lf (n, _) -> n
+      | Mt (n, _) -> n
+
+    let rec add_leaf_to_mtree paths leaf_value = function
+      | Lf _ -> raise (Failure "Can't add a leaf on a leaf")
+      | Mt (root, children) ->
+        (match paths with
+         | [] ->
+           (match children with
+            | [] -> Lf(root, leaf_value)
+            | _ -> raise (Failure "Can't add a leaf on a tree node"))
+         | node :: rest_paths ->
+           try 
+             let t = List.find (has_name node) children in
+             (match t with
+              | Lf (_, _) -> raise (Failure "Can't overwrite an existing leaf")
+              | Mt (node, children_of_node) ->
+                let mt = add_leaf_to_mtree rest_paths leaf_value (Mt(node, children_of_node)) in
+                Mt(root, mt :: (List.filter (fun n -> not (has_name node n)) children)))
+           with Not_found -> 
+             Mt(root, (add_leaf_to_mtree rest_paths leaf_value (Mt(node, []))) :: children))
+
+    let construct_mtree mtree (path, leaf_value) =
+      let open Stdext.Xstringext in
+      let nodes = String.split_f (fun s -> s = '/') path in
+      add_leaf_to_mtree nodes leaf_value mtree
+
+    let rec list_helper children = function
+      | [] -> List.map get_name children
+      | node :: rest_paths ->
+        try
+          match List.find (has_name node) children with
+          | Lf (_, _) -> []
+          | Mt (_, children_of_node) -> list_helper children_of_node rest_paths
+        with Not_found -> []
+
+    let list mtree path =
+      match mtree with
+      | Lf (_, _) -> []
+      | Mt (_, children) ->
+        let open Stdext.Xstringext in
+        let paths = String.split_f (fun s -> s = '/') path in
+        list_helper children paths
+
+    let rec lookup_helper mtree = function
+      | [] -> 
+        (match mtree with
+         | Lf (_, v) -> Some v
+         | Mt (_, _) -> None)
+      | node :: rest_paths ->
+        (match mtree with
+         | Lf (l, v) -> lookup_helper (Lf(l, v)) rest_paths
+         | Mt (_, children) -> 
+           try
+             lookup_helper (List.find (has_name node) children) rest_paths
+           with Not_found -> None)
+
+    let lookup mtree path =
+        let open Stdext.Xstringext in
+        let paths = String.split_f (fun s -> s = '/') path in
+        lookup_helper mtree paths
+
+
+    let transform input = 
+      let tree = List.fold_left construct_mtree (Mt("", [])) input in
+      let guest_metrics = Xapi_guest_agent.get_initial_guest_metrics (lookup tree) (list tree) in
+      guest_metrics.Xapi_guest_agent.networks
+
+
+    let tests = [
+      (* basic cases *)
+      [ "attr/vif/0/ipv6/0", "192.168.0.1";
```

Comment:
Oh, and that does not look like an IPv6 address... (doesn't matter for the test, but still ;) )